### ISSUE TITLE
Clarifying data-loading needs.

### ISSUE THREAD

**Issue by AndrewIngram** [ReactTraining/react-router-addons-routes#2]
I think specifically calling out server-side rendering as reasons for needing a data loading strategy is a bit misleading, because client-side has the same requirements.

First let's assume we don't do server-side rendering. We want the first render on the client to complete as quickly as possible, so we'd often want to use the same strategy we'd use on the server on the client instead. I say "often" because there are scenarios where you'd be happy to defer loading and show a spinner or something instead, though my approach to that has always been to just have a separate loading strategy for deferred content.

So regardless of whether you do it on client or server, the first render data requirements are the same. But hopefully if you do it on the server, you include a data payload in the markup that means you don't do all the work again on the client.

The second one is data loading on transitions. Assuming our traversal takes us more than one hop in our hierarchy of "entry points", we're going to hit the same problem of request waterfalls that we've solved for the initial page load. So if I go from A/B/C to A/D/E, I need to efficiently load both D and E (i.e. in parallel or as one request), or if I go from A/B/C to A/B/F/G, I need to efficiently load F and G.

Now, the same solution of matching on statically defined routes is going to partially solve the problem, but what we also need is an idiomatic way (or at least an example) of blocking the transition (or at least the render) until the new data requirements are satisfied.

**Comment by ryanflorence** [comment]
Wrote the thing up real quick, but it does say this :)

> This can be used server-side or in a data component's lifecycle to determine which routes are going to be rendered next.

Are you looking for me to write an example or just adjust the wording?

**Comment by ryanflorence** [comment]
I'd love to see somebody port my old https://github.com/ryanflorence/async-props to this. it handles all of these transitions.

**Comment by ryanflorence** [comment]
Sorry for the multiple messages, I was on my phone.

Here's how the "render the old screen" strategy:

Save old props and re-render before loading, sending down a "loading" prop so the UI can indicate it: https://github.com/ryanflorence/async-props/blob/master/modules/AsyncProps.js#L257-L260

Decide which props to use to render, the current ones or the previous ones: https://github.com/ryanflorence/async-props/blob/master/modules/AsyncProps.js#L300-L301

Finally, get rid of the old props: https://github.com/ryanflorence/async-props/blob/master/modules/AsyncProps.js#L279-L283

**Comment by mjackson** [comment]
What do you want to be done here @AndrewIngram?

**Comment by AndrewIngram** [comment]
I've been super swamped the last 2 weeks, but hopefully i'll be able to put some time into trying out Relay integration this weekend, which should help clarify the situation.

**Comment by tomatau** [comment]
After exploring many universal data fetching implementations. I found the API for https://github.com/markdalgleish/redial to be the most declarative and compose-able (sorry ryanflorence)

Async props seems to take too many concerns with it's script tag approach to providing the prefetched data instead of just providing the raw data.

`matchRoutesToLocation` would work perfectly well with a redial style approach. Infact I don't think redial would even need porting as it stands, it would just work given the array of route components to trigger.

**Comment by bsideup** [comment]
Take a look at https://bsideup.gitbooks.io/rxconnect/content/ as well:)
